week-2/mandatory/3-api/server.js

```diff
@@ -0,0 +1,382 @@
+      "values ($1,$2,$3)";
+  if (price > 0) {
+    db.query(sql_supp, [supplier], (error, result) => {
+      if (error) {
+        console.log("pg error code:", error.code);
+        res.status(500).json(serverError);
+      } else if (result.rows.length > 0) {
+        db.query(sql_prod, [product], (error, result) => {
+          if (error) {
+            console.log("pg error code:", error.code);
+            res.status(500).json(serverError);
+          } else if (result.rows.length > 0) {
+            db.query(sql_insert, [product, supplier, price], (error) => {
+              if (error) {
+                console.log("pg error code:", error.code);
+                res.status(500).json(serverError);
+              } else {
+                res.status(201).json("Product price added.");
+              }
+            });
+          } else {
+            res.status(404).json("Can't find a product with this ID");
+          }
+        });
+      } else {
+        res.status(404).json("Can't find a supplier with this ID");
+      }
+    });
+  } else {
+    res.status(400).json("The price needs to be larger than 0");
+  }
+});
+
+app.post("/customers/:customerId/orders", (req, res) => {
+  let id = parseInt(req.params.customerId),
+    sql_orders =
+      "insert into orders (order_date, order_reference, customer_id)" +
+      "values (current_date,$1,$2) returning order_reference",
+    sql_customers = "select * from customers where id = $1",
+    sql_orderNumber =
+      "select order_reference from orders order by id DESC limit 1";
+
+  db.query(sql_customers, [id], (error, result) => {
+    if (error) {
+      console.log("pg error code:", error.code);
+      res.status(500).json(serverError);
+    } else if (result.rows.length > 0) {
+      db.query(sql_orderNumber, (error, result) => {
+        if (error) {
+          console.log("pg error code:", error.code);
+          res.status(500).json(serverError);
+        } else {
+          let lastOrder = result.rows[0].order_reference;
+          //orderRef functions are borrowed from https://electrictoolbox.com/pad-number-zeroes-javascript/ and SOF site
+          let orderRef = lastOrder.replace(/(\d+)+/g, function (match, number) {
+            function pad(number, length) {
+              let str = "" + number;
+              while (str.length < length) {
+                str = "0" + str;
+              }
+              return str;
+            }
+            return pad(parseInt(number) + 1, 3);
+          });
+          db.query(sql_orders, [orderRef, id], (error, result) => {
+            if (error) {
+              console.log("pg error code:", error.code);
+              res.status(500).json(serverError);
+            } else {
+              res
+                .status(201)
+                .json(
+                  `Order placed with reference ${result.rows[0].order_reference}`
+                );
+            }
+          });
+        }
+      });
+    }
+  });
+});
```

Comment:
The model answer we have been given assumes that the user provides the order reference, but I can see why you have assumed the user would not know this. You have come up with a good solution to this, and referenced where you found the code, which is good practice for substantial pieces of copied code..